### PR TITLE
[PAN-2949] Add rewind to block functionality

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Retesteth will often request an in-process blockchain be rewound one block so
that it can attempt the same test with tweaked parameters. Add support for this.

### CODE REVIEW

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchainTest.java [hunk 1]
```diff
@@ -455,6 +455,44 @@ public void reorgWithOverlappingTransactions() {
...
+    for (int i = 1; i < chain.size(); i++) {
+      blockchain.appendBlock(chain.get(i), blockReceipts.get(i));
+    }
+    final Block originalHead = blockchain.getChainHeadBlock();
+    final Block targetHead = blockchain.getBlockByHash(originalHead.getHeader().getParentHash());
+
+    // rewind it by 1 block
+    blockchain.rewindToBlock(targetHead.getHeader().getNumber());
+
+    // Check chain has not reorganized
```
Comment (mbaxter) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchainTest.java]: This comment looks wrong :)

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java [hunk 2]
```diff
@@ -362,6 +362,34 @@ private BlockAddedEvent handleChainReorg(
...
+      return false;
+    }
+    final Block block = new Block(oldBlockHeader.get(), oldBlockBody.get());
+
+    final BlockchainStorage.Updater updater = blockchainStorage.updater();
+    final BlockAddedEvent result = this.handleChainReorg(updater, block);
+    updater.commit();
+
+    if (result.isNewCanonicalHead()) {
+      chainHeader = block.getHeader();
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java]: Done.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java [hunk 3]
```diff
@@ -362,6 +362,34 @@ private BlockAddedEvent handleChainReorg(
...
+    if (oldBlockHeader.isEmpty() || oldBlockBody.isEmpty()) {
+      return false;
+    }
+    final Block block = new Block(oldBlockHeader.get(), oldBlockBody.get());
+
+    final BlockchainStorage.Updater updater = blockchainStorage.updater();
+    final BlockAddedEvent result = this.handleChainReorg(updater, block);
+    updater.commit();
+
+    if (result.isNewCanonicalHead()) {
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java]: Done.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java [hunk 4]
```diff
@@ -362,6 +362,34 @@ private BlockAddedEvent handleChainReorg(
...
   }
 
+  public boolean rewindToBlock(final long blockNumber) {
+    final Optional<Hash> blockHash = blockchainStorage.getBlockHash(blockNumber);
+    if (blockHash.isEmpty()) {
+      return false;
+    }
+
+    final Optional<BlockHeader> oldBlockHeader = blockchainStorage.getBlockHeader(blockHash.get());
+    final Optional<BlockBody> oldBlockBody = blockchainStorage.getBlockBody(blockHash.get());
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java]: Done.

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchainTest.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchainTest.java]: reworded.

**Pull request merged**